src/wiki/plugins/links/mdx/urlize.py

```diff
@@ -2,98 +2,138 @@
+#   href).
+# - protocol (protocol, together with the trailing ://)
+# - host (just the host part)
+# - port (just the port number)
+# - path (path, combined with any additional GET parameters)
+# - end (character denoting end of URL)
+#
+URLIZE_RE = (
+    r'(?P<begin>[\(\<])?(?P<url>'  # begin url group
 
->>> md.convert('go to <http://example.com> now!')
-'<p>go to <a href="http://example.com">http://example.com</a> now!</p>'
+    # Leading protocol specification.
+    r'(?P<protocol>http://|https://|ftp://|ftps://|)'
 
-Negative examples:
+    # Host identifier
+    r'(?P<host>'  # begin host identifier group
 
->>> md.convert('del.icio.us')
-'<p>del.icio.us</p>'
+    r'[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}|'  # IPv4, match before FQDN
+    r'\[?([A-F0-9]{1,4}:){7}([A-F0-9]{1,4})\]?|'  # IPv6, full form
+    r'\[?:(:[A-F0-9]{1,4}){1,6}\]?|'  # IPv6, leading zeros removed
+    r'([A-F0-9]{1,4}:){1,6}:([A-F0-9]{1,4}){1,6}|'  # IPv6, zeros in middle removed.
+    r'\[?([A-F0-9]{1,4}:){1,6}:\]?|'  # IPv6, trailing zeros removed
+    r'\[?::\]?|'  # IPv6, just "empty" address
+    r'([A-Z0-9]([A-Z0-9-]{0,61}[A-Z0-9])?\.)+([A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # FQDN
+    r'localhost'  # localhost
+    r')'  # end host identifier group
 
-"""
+    # Optional port
+    r'(:(?P<port>[0-9]+))?'
 
+    # Optional trailing slash with path and GET parameters.
+    r'(/(?P<path>[^\s\[\(\]\)\<\>]*))?'
 
-# Taken from Django trunk 2f121dfe635b3f497fe1fe03bc8eb97cdf5083b3
-# https://github.com/django/django/blob/master/django/core/validators.py#L47
-URLIZE_RE = (
-    r'((?:(?:http|ftp)s?://|www\.)'  # http:// or https://
-    # domain...
-    r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'
-    r'localhost|'  # localhost...
-    r'[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}|'  # ...or ipv4
-    r'\[?[A-F0-9]*:[A-F0-9:]+\]?)'  # ...or ipv6
-    r'(?::[0-9]+)?'  # optional port
-    r'(?:/[^\s\[\(\]\)]*(?:\s+|$))?)'
+    r')(?P<end>[\)\>])?'  # end url group
 )
 
 
 class UrlizePattern(markdown.inlinepatterns.Pattern):
 
-    def __init__(self, pattern, markdown_instance=None):
-        markdown.inlinepatterns.Pattern.__init__(
-            self,
-            pattern,
-            markdown_instance=markdown_instance)
-        self.compiled_re = re.compile("^(.*?)%s(.*?)$" % pattern,
-                                      re.DOTALL | re.UNICODE | re.IGNORECASE)
+    def getCompiledRegExp(self):
+        """
+        Return compiled regular expression for matching the URL
+        patterns. We introduce case-insensitive matching in addition
+        to standard matching flags added by parent class.
+        """
 
-    """ Return a link Element given an autolink (`http://example/com`). """
+        # Ensure links are matched only if they stand on their own to avoid bad matches etc.
+        return re.compile(r'^(|.*?\s)%s(\s.*?|)$' % URLIZE_RE, re.DOTALL | re.UNICODE | re.IGNORECASE)
```

Comment:
There's already the group called `end` in `URLIZE_RE`, perhaps somehow merge the remaining regular expression logic with that and add a group `start` ?

Reply:
I think begin/end will end-up being dropped once I remove the `()` and `<>` syntax support.